// @caido/eslint-config@0.5.0/dist/index.js
// src/plugins.ts
import { default as default2 } from "@eslint/js";
import { default as default3 } from "eslint-plugin-vue";
import { default as default4 } from "typescript-eslint";
import { default as default5 } from "eslint-plugin-prettier/recommended";
import { default as default6 } from "eslint-plugin-compat";
import { default as default7 } from "eslint-plugin-import";
import { default as default8 } from "eslint-plugin-n";
import { default as default9 } from "eslint-config-prettier/flat";

// src/configs/javascript.ts
var javascript = () => {
  return [
    default2.configs.recommended,
    {
      name: "caido/javascript",
      rules: {
        eqeqeq: "error",
        "no-empty-pattern": "off",
        "sort-imports": [
          "warn",
          {
            ignoreCase: true,
            ignoreDeclarationSort: true
          }
        ]
      }
    }
  ];
};

// src/globs.ts
var GLOB_VUE = "**/*.vue";
var GLOB_TS = "**/*.ts";

// src/configs/typescript.ts
var typescript = (options) => {
  return [
    ...default4.configs.recommendedTypeChecked,
    {
      name: "caido/typescript",
      files: [GLOB_TS, ...options.extraFiles.map((file) => `**/*.${file}`)],
      languageOptions: {
        parserOptions: {
          projectService: true,
          extraFileExtensions: [
            ...options.extraFiles.map((file) => `.${file}`)
          ]
        }
      },
      rules: {
        // Disabled
        "@typescript-eslint/no-unsafe-enum-comparison": "off",
        "@typescript-eslint/ban-ts-ignore": "off",
        "@typescript-eslint/explicit-function-return-type": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "@typescript-eslint/no-unsafe-argument": "off",
        "@typescript-eslint/no-unsafe-member-access": "off",
        "@typescript-eslint/no-unsafe-call": "off",
        "@typescript-eslint/no-floating-promises": "off",
        "@typescript-eslint/no-unsafe-assignment": "off",
        "@typescript-eslint/restrict-template-expressions": "off",
        "@typescript-eslint/no-unsafe-return": "off",
        "@typescript-eslint/no-implied-eval": "off",
        "@typescript-eslint/unbound-method": "off",
        "@typescript-eslint/no-unused-expressions": "off",
        "@typescript-eslint/no-misused-promises": "off",
        "@typescript-eslint/strict-boolean-expressions": "error",
        "@typescript-eslint/consistent-type-imports": [
          "error",
          {
            fixStyle: "inline-type-imports"
          }
        ],
        "@typescript-eslint/switch-exhaustiveness-check": [
          "error",
          {
            considerDefaultExhaustiveForUnions: true
          }
        ],
        // Disable no-unused-vars and uses noUnusedLocals: true in tsconfig.json instead
        // Reference: https://github.com/johnsoncodehk/volar/issues/47
        "@typescript-eslint/no-unused-vars": "off",
        "@typescript-eslint/no-restricted-types": [
          "error",
          {
            types: {
              null: "Use 'undefined' instead of 'null'"
            }
          }
        ]
      }
    }
  ];
};

// src/configs/vue.ts
import parserVue from "vue-eslint-parser";
var vue = (options) => {
  return [
    ...default3.configs["flat/recommended"].map((config) => {
      return {
        ...config,
        files: [GLOB_VUE]
      };
    }),
    {
      name: "caido/vue",
      files: [GLOB_VUE],
      rules: {
        // Disabled because clashes with prettier
        "vue/html-closing-bracket-newline": "off",
        "vue/multi-word-component-names": "off",
        "vue/require-v-for-key": "off",
        "vue/no-template-shadow": "off",
        "vue/valid-v-for": "off",
        "vue/max-attributes-per-line": "off",
        // Necessary configuration due to conflict with eslint-config-prettier
        // https://github.com/prettier/eslint-config-prettier?tab=readme-ov-file#vuehtml-self-closing
        "vue/html-self-closing": [
          "error",
          {
            html: {
              void: "any"
            }
          }
        ]
      },
      languageOptions: {
        parser: parserVue,
        parserOptions: {
          ecmaVersion: "latest",
          // Use different parsers for different languages
          // We do this to improve performance
          // https://github.com/vuejs/vue-eslint-parser/issues/104#issuecomment-875149361
          parser: {
            ts: default4.parser,
            js: default4.parser,
            "<template>": "espree"
          },
          sourceType: "module",
          extraFileExtensions: [
            ...options.extraFiles.map((file) => `.${file}`)
          ]
        }
      }
    }
  ];
};

// src/configs/imports.ts
var imports = () => {
  return [
    default7.flatConfigs["recommended"],
    {
      name: "import/typescript",
      ...default7.flatConfigs["typescript"]
    },
    {
      name: "caido/imports",
      rules: {
        // Disabled for performance issues
        // Reference: https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting/#eslint-plugin-import
        "import/namespace": "off",
        "import/first": "error",
        // Disabled since this is already handled by Typescript
        "import/no-unresolved": "off",
        "import/order": [
          "warn",
          {
            "newlines-between": "always",
            warnOnUnassignedImports: true,
            groups: [
              "builtin",
              "external",
              "internal",
              "parent",
              "sibling",
              "index"
            ],
            alphabetize: {
              order: "asc",
              caseInsensitive: true
            }
          }
        ]
      }
    }
  ];
};

// src/configs/compat.ts
var compat = () => {
  return [default6.configs["flat/recommended"]];
};

// src/configs/prettier.ts
var prettier = () => {
  return [
    {
      name: "prettier/recommended",
      ...default5
    },
    {
      ...default9,
      name: "config-prettier/recommended"
    }
  ];
};

// src/configs/ignore.ts
var ignore = () => {
  return [
    {
      name: "caido/ignore",
      ignores: ["**/coverage", "**/dist"]
    }
  ];
};

// src/configs/node.ts
var node = () => {
  return [
    {
      plugins: {
        n: default8
      },
      rules: {
        "n/no-unsupported-features/es-builtins": "error",
        "n/no-unsupported-features/es-syntax": "error",
        "n/no-unsupported-features/node-builtins": "error"
      }
    }
  ];
};

// src/index.ts
var DEFAULT_OPTIONS = {
  javascript: true,
  typescript: true,
  vue: true,
  compat: true,
  imports: true,
  prettier: true,
  node: false
};
var defaultConfig = (options = {}) => {
  const finalOptions = {
    ...DEFAULT_OPTIONS,
    ...options
  };
  const extraFilesOptions = {
    extraFiles: []
  };
  if (finalOptions.typescript) {
    extraFilesOptions.extraFiles.push("vue");
    extraFilesOptions.extraFiles.push("tsx");
  }
  return [
    ...ignore(),
    ...finalOptions.javascript ? javascript() : [],
    ...finalOptions.typescript ? typescript(extraFilesOptions) : [],
    ...finalOptions.vue ? vue(extraFilesOptions) : [],
    ...finalOptions.imports ? imports() : [],
    ...finalOptions.compat ? compat() : [],
    ...finalOptions.prettier ? prettier() : [],
    ...finalOptions.node ? node() : []
  ];
};
export {
  defaultConfig
};
